### DISCUSSION TITLE
Adapting NaN as NULL

### DISCUSSION THREAD

**Discussion by thrasibule** [psycopg/psycopg#240] [Q&A]
Is there a way to insert NaN floats as NULL? I tried the following Dumper but that doesn't work

```python
class NaNAsNULLFloatDumper(FloatDumper):
    def dump(self, obj):
        if math.isnan(obj):
            return b"'NULL'::float8"
        else:
            return super().dump(obj)
```

**Comment by dvarrazzo** [comment]
At the moment there isn't one on the main query path. Inserting a null takes a different path than insetting any non-null value.

What you are trying to do maybe can be achieved implementing the method `quote()` of the dumper - see numeric or bool dumpers for instances. But in order to make use of it you will need to bind the query client side by using `psycopg.sql` objects such as `SQL` and `Literal`.

For the future I'm thinking about a cursor subclass to bind parameters cient-side (see #101, #205), but real use cases are struggling to emerge. Even yours is not really a transformation that makes sense: NaN and NULL are hardly two overlapping concepts.

**Reply by thrasibule** [reply]
Thanks for the explanation. I was hoping using a string NULL would work. The use case I have is pandas. Pandas uses NaN as a sentinel to mark missing values for floats (and R as well for that matter). If I understand correctly the decision to send NULL is done before any python value adaptation, that's why there is no way to hook into that decision? I guess I'll just replace NaN with None client side.
